x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/privileges/UserPrivilegeResolver.java

```diff
@@ -128,18 +133,34 @@ private void buildResourcePrivilege(
         ServiceProviderPrivileges service,
         ActionListener<RoleDescriptor.ApplicationResourcePrivileges> listener
     ) {
-        actionsResolver.getActions(service.getApplicationName(), listener.delegateFailureAndWrap((delegate, actions) -> {
-            if (actions == null || actions.isEmpty()) {
-                logger.warn("No application-privilege actions defined for application [{}]", service.getApplicationName());
-                delegate.onResponse(null);
-            } else {
-                logger.debug("Using actions [{}] for application [{}]", actions, service.getApplicationName());
-                final RoleDescriptor.ApplicationResourcePrivileges.Builder builder = RoleDescriptor.ApplicationResourcePrivileges.builder();
-                builder.application(service.getApplicationName());
-                builder.resources(service.getResource());
-                builder.privileges(actions);
-                delegate.onResponse(builder.build());
-            }
+        // We need to enumerate possible actions that might be authorized for the user. Here we combine actions that
+        // have been granted to the user via roles and other actions that are registered privileges for the given
+        // application. These actions will be checked by a has-privileges check above
+        final GetUserPrivilegesRequest request = new GetUserPrivilegesRequestBuilder(client).username(securityContext.getUser().principal())
+            .request();
+        client.execute(GetUserPrivilegesAction.INSTANCE, request, listener.delegateFailureAndWrap((delegate, userPrivileges) -> {
+            Set<String> actionsFromRoles = userPrivileges.getApplicationPrivileges()
+                .stream()
+                .filter(appPriv -> appPriv.getApplication().equals(service.getApplicationName()))
+                .map(appPriv -> appPriv.getPrivileges())
+                .flatMap(Arrays::stream)
+                .collect(Collectors.toUnmodifiableSet());
+            actionsResolver.getActions(service.getApplicationName(), delegate.delegateFailureAndWrap((delegate2, appDefinedActions) -> {
```

Comment:
Reviewers, is there a standard way to do two async tasks in parallel? I could fetch both `userPrivileges` and `appDefinedActions` concurrently instead of sequentially.

Reply:
Generally we don't both tying to run them in parallel. Most of the time we're aiming for throughput rather than latency, and are happy to run them sequentially.

You could use `GroupedActionListener<Set<String>>` for this.
I'll paste one such option it as a suggestion below.